Sync with upstream 2021.12.20

nunnatsa opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release notes**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Sync with kubernetes-sigs/cluster-api-provider-kubevirt, 2021.12.20
```

rgolangh commented:
@nirarg @bardielle PTAL

rgolangh commented:
This should fix the images job for the CI PR https://github.com/openshift/release/pull/12318

bardielle reviewed (approved):
/lgtm

nirarg commented:
/lgtm

nirarg commented:
@chenyosef Please merge

davidvossel committed:
Only inject ssh keys when cloud-config user data is used

Signed-off-by: David Vossel <davidvossel@gmail.com>

davidvossel committed:
Update machine unit test to require ready condition for detecting is booted

Signed-off-by: David Vossel <davidvossel@gmail.com>

davidvossel committed:
Removes spec.Bootstrapped in favor of conditions

Signed-off-by: David Vossel <davidvossel@gmail.com>

davidvossel committed:
Ensure that bootstrap exec condition is set during functional test

Signed-off-by: David Vossel <davidvossel@gmail.com>

davidvossel committed:
Add unit tests for determining if boostrap detection is possible

Signed-off-by: David Vossel <davidvossel@gmail.com>

davidvossel committed:
Add unit test for detecting cloud-config user data type

Signed-off-by: David Vossel <davidvossel@gmail.com>

rmohr committed:
Build and test the code with github actions

Signed-off-by: Roman Mohr <rmohr@redhat.com>

rmohr committed:
Add e2e test lane

Signed-off-by: Roman Mohr <rmohr@redhat.com>

agradouski committed:
Support for external infra clusters:

- add InfraClusterSecretRef to KubevirtClusterSpec
- add example template
- add InfraCluster construct to incapsulate infra cluster client and context
- move cluster management resource cleanup logic to `reconcileDelete` func in controllers
- enhance logging

Signed-off-by: Alex Gradouski <agradouski@apple.com>

aaseem committed:
Update README.md

aaseem committed:
Meeting details updated

davidvossel committed:
Change IsBooted to IsReady in order to better reflect what is being checked

Signed-off-by: David Vossel <davidvossel@gmail.com>

agradouski committed:
Support for external infra clusters:

- return on errors, with retries

Signed-off-by: Alex Gradouski <agradouski@apple.com>

k8s-ci-robot committed:
Merge pull request #30 from davidvossel/cloud-init-fix

Make ssh bootstrap detection optional

1abhisheksarkar committed:
Update README.md

davidvossel committed:
kubevirt machine reconcile normal unit tests

Signed-off-by: David Vossel <davidvossel@gmail.com>

davidvossel committed:
remove race condition when searching for vmi immediately after vm creation

Signed-off-by: David Vossel <davidvossel@gmail.com>

k8s-ci-robot committed:
Merge pull request #43 from TheAbhishekS/abhishek1

Update README.md

cchengleo committed:
Add dockfile to build manager binary and emit production image

k8s-ci-robot committed:
Merge pull request #38 from rmohr/testing

Add unit- and functional test runners based on github actions

PranshuSrivastava committed:
added meeting info to the README

k8s-ci-robot committed:
Merge pull request #48 from PranshuSrivastava/meeting

added meeting info to the README

rmohr committed:
Make secrets for third-party PRs available

If a PR gets labeled with `ok-to-test` the integration test job is
triggered and this github action has access to secrets.

The github action does for security resaons not allow secrets on all
PRs, so we add this extra step.

Still, the action itself, while using a secret, does not have any
interaction points with users. The PR content is actually not even
checked out.

Signed-off-by: Roman Mohr <rmohr@redhat.com>

k8s-ci-robot committed:
Merge pull request #49 from rmohr/make-secrets-available

Make secrets for third-party PRs available

k8s-ci-robot committed:
Merge pull request #46 from cchengleo/update-dockerfile

Add Dockerfile to build manager binary and emit production image

nirarg committed:
Change OWNERS file

k8s-ci-robot committed:
Merge pull request #44 from davidvossel/reconcile-tests-1

Fixes "failed to create VM instance" error

aaseem committed:
Merge branch 'kubernetes-sigs:main' into main

aaseem committed:
Update README.md

rmohr committed:
Make integration tests independent of go version changes

With the new Dockerfile in place which builds the manager binary, the
binary does not have to be built upfront with the external golang
version. Instead the internal one can be used which makes for instance
integration jobs independent of the go versions installed there.

Signed-off-by: Roman Mohr <rmohr@redhat.com>

rmohr committed:
Give github actions better names

Signed-off-by: Roman Mohr <rmohr@redhat.com>

agradouski committed:
rebase

Signed-off-by: Alex Gradouski <agradouski@apple.com>

agradouski committed:
Merge branch 'main' into main

k8s-ci-robot committed:
Merge pull request #53 from rmohr/integration-tests-no-external-golang

Make integration tests independent of go version changes

k8s-ci-robot committed:
Merge pull request #52 from aaseem/main

Meetings Details update in README.md

agradouski committed:
fix unit tests

Signed-off-by: Alex Gradouski <agradouski@apple.com>

k8s-ci-robot committed:
Merge pull request #39 from agradouski/main

Support for external infra clusters

rmohr commented on .github/workflows/e2e.yaml:
```diff
@@ -0,0 +1,25 @@
+          repository: kubevirt/project-infra
+          path: project-infra
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
```
Will be interesting to see if github actions will be used here too. We potentially have to delete them.

nunnatsa replied:
Easy to prevent by adding 
```
if: (github.repository == 'kubernetes-sigs/cluster-api-provider-kubevirt')
```

in the kubernetes-sigs repo

rmohr reviewed (commented):
/lgtm
